### PR TITLE
Refactor out sources logic into separate package

### PR DESCRIPTION
**Pull request by eseliger**
We had syncer, reconciler executor and the service all three require a changeset source, with some sort of authentication in place. The logic for this was different in all three places, and I was about to introduce a 4th place where I'd use it - so I figured this is a good moment to do a nice refactor and have a single source of truth for external service selection, authenticator usage and clone URL computation. Say hello to the new `sources` package. We might be able to wrap existing `GitHubSource`,... and such in the future so we can pull the `ChangesetSource` logic into our codebase ... let see, stay tuned. For now, this is enabling us to easily get an authenticated `ChangesetSource` for a given changeset or repo.

### CODE REVIEW

File: enterprise/internal/batches/sources/sources.go [hunk 1]
```diff
@@ -0,0 +1,433 @@
...
+		IDs: extSvcIDs,
+	})
+}
+
+// FromRepoSource returns a BatchesSource for a given repos.Source.
+func (s *Sourcer) FromRepoSource(src repos.Source) (*BatchesSource, error) {
+	return batchesSourceFromRepoSource(src, s.store)
+}
+
+func (s *Sourcer) loadBatchesSource(ctx context.Context, opts database.ExternalServicesListOptions) (*BatchesSource, error) {
```
Comment (LawnGnome) [inline enterprise/internal/batches/sources/sources.go:114]: Does this need to take a full `database.ExternalServicesListOptions` (and be tied to more of an implementation detail of that package) when the only callers provide a slice of IDs?

File: enterprise/internal/batches/sources/sources.go [hunk 2]
```diff
@@ -0,0 +1,433 @@
...
+	if err != nil && !errcode.IsNotFound(err) {
+		return nil, err
+	}
+	if cred != nil {
+		return cred.Credential, nil
+	}
+	return nil, nil
+}
+
+// v attempts to find a site credential for the given repo.
```
Comment (LawnGnome) [inline enterprise/internal/batches/sources/sources.go:371]:
```suggestion
// loadSiteCredential attempts to find a site credential for the given repo.
```

File: enterprise/internal/batches/reconciler/executor.go [hunk 3]
```diff
@@ -459,6 +342,50 @@ func (e *executor) sleep() {
...
+				return nil, &errMissingCredentials{repo: string(repo.Name)}
+			case sources.ErrNoSSHCredential:
+				return nil, &errNoSSHCredential{}
+			default:
+				if enpc, ok := err.(sources.ErrNoPushCredentials); ok {
+					return nil, &errNoPushCredentials{credentialsType: enpc.CredentialsType}
+				}
+				return nil, err
+			}
+		}
```
Comment (LawnGnome) [inline enterprise/internal/batches/reconciler/executor.go:373]: This error marshalling feels a little weird to me: could the callers that need to introspect on the error return value do so on the original `sources` errors instead?

File: enterprise/internal/batches/resolvers/main_test.go [hunk 4]
```diff
@@ -255,3 +256,16 @@ func pruneSiteCredentials(t *testing.T, cstore *store.Store) {
...
+		return &encryption.RSAKey{
+			PrivateKey: "private",
+			Passphrase: "pass",
+			PublicKey:  "public",
+		}, nil
+	}
+	t.Cleanup(func() {
+		encryption.MockGenerateRSAKey = nil
+	})
+}
```
Comment (LawnGnome) [inline enterprise/internal/batches/resolvers/main_test.go:271]: I'm glad you added this (and the plumbing below it): there's no need for our test runners to burn CPU cycles creating real SSH keys we'll never use. :evergreen_tree:

**Review by LawnGnome — APPROVED** [review]
This is a really useful simplification: if you exclude the new unit tests, this is a decent net reduction in code, which is usually a pretty sure sign of a good refactor.

Nice work!

File: enterprise/internal/batches/reconciler/executor.go (hunk 3 above)
Comment (eseliger) [reply enterprise/internal/batches/reconciler/executor.go:373]: These errors fulfil the `NonRetryable` interface, which is sorta an implementation detail of the executor/worker thingy, which I didn't want to bleed through all the layers, but if you think this is ugly I can certainly change it back to be on the errors thrown by the source directly

Comment (LawnGnome) [reply enterprise/internal/batches/reconciler/executor.go:373]: No, I missed that. Fair enough, then. (I mean, it's a _bit_ ugly even knowing that, but I'm going to blame Go and move on with my life.)

**Pull request merged**